### PR TITLE
Triangulated fill outlines

### PR DESCRIPTION
**Pull request by stefankarschti**
feat: Fill layer to generate outlines composed of triangle primitives
- Created a new addition to the fill generator functions to generate this new fill+outline variant
- The Fill bucket generates fill outlines with triangle primitives, compatible with the Line shader
- The Fill bucket uses `gfx::generateFillAndOutineBuffers` to generate its buffers
- Added an option to enable/disable this feature at compile time with the `MLN_TRIANGULATE_FILL_OUTLINES` #define in `fill_bucket.hpp`

Refs: #1768

### CODE REVIEW

**Comment by stefankarschti** [comment]
WIP
image

**Comment by stefankarschti** [comment]
Comparison of fill outlines

_Legacy renderer:_
IMG_0032

_Drawable renderer:_
IMG_0031

**Comment by stefankarschti** [comment]
_Fill pattern outline drawn with the legacy renderer:_
IMG_0033-legacy

_And with the Metal renderer (no outline):_
IMG_0034-metal

**Comment by stefankarschti** [comment]
performance impact versus [main@a73e88fee28593a0e66be1e67d37ae444b0bfc6c ](https://github.com/maplibre/maplibre-native/commit/a73e88fee28593a0e66be1e67d37ae444b0bfc6c) 
Screenshot 2023-11-06 at 22 13 30

File: platform/ios/src/MLNMapView+Metal.mm [hunk 1]
```diff
@@ -82,7 +82,7 @@ void swap() override {
 
         // Un-comment for synchronous, which can help troubleshoot rendering problems,
         // particularly those related to resource tracking and multiple queued buffers.
-        //[commandBuffer waitUntilCompleted];
+        [commandBuffer waitUntilCompleted];
```
Comment (sjg-wdw) [inline platform/ios/src/MLNMapView+Metal.mm:85]: Probably want to turn this back off.
It might be changing the frame rendering time too.

Comment (stefankarschti) [reply platform/ios/src/MLNMapView+Metal.mm:85]: oops. correct. remeasuring.

**Comment by stefankarschti** [comment]
new measurements after commenting out `[commandBuffer waitUntilCompleted];`
image

**Comment by stefankarschti** [comment]
New timing measured by averaging 3 runs on _iPad 5th gen_.
In _LineBasicShader v1_ the `interpolation` UBO is deleted.
In _LineBasicShader v2_ the `blur`, `gap width`, `offset` UBO fields are removed; `permutation` and `expression` UBOs are also removed on Metal.

image

**Comment by sjg-wdw** [comment]
Looks like reasonable numbers.  
One more useful measurement might be lines with the single pixel Metal implementation (e.g. what Metal provides).  Just to see what lines are costing us.

**Comment by stefankarschti** [comment]
> Looks like reasonable numbers. One more useful measurement might be lines with the single pixel Metal implementation (e.g. what Metal provides). Just to see what lines are costing us.

This is what metal on `main` has, 2 pixel wide lines (which render as 1 pixel)...

**Comment by sjg-wdw** [comment]
Hmmm, still some cost to the simpler version.

**Comment by sjg-wdw** [comment]
Staring at the numbers a bit more.... I assume they'll be slightly different on more recent hardware, but they're unlikely to break the pattern we've seen before.

I think we want to go with the v2 shader and we may also want to provide an option to turn that off and use Metal lines.
But after we've got the v2 shader merged in, we should see how it fares in a Metal frame analysis.  Maybe those tools can see something we can't.

**Review by louwers — COMMENTED** [review]
Why is the `MLN_TRIANGULATE_FILL_OUTLINES` flag needed? Is it only because the legacy renderer does not support it? In this case, please add a comment to the issue to remove the legacy renderer to remove this build flag as well.

In general, I think we should avoid build time configuration as much as possible because only the default is tested / compiled on CI.

**Comment by stefankarschti** [comment]
@louwers : I added that flag following Steve's recommendation to "provide an option to turn that off and use Metal lines"...
CC @sjg-wdw

**Comment by sjg-wdw** [comment]
> @louwers : I added that flag following Steve's recommendation to "provide an option to turn that off and use Metal lines"... CC @sjg-wdw

Nice!

**Comment by alexcristici** [comment]
With `#define MLN_TRIANGULATE_FILL_OUTLINES 1` render test `render-tests/fill-outline-color/property-function` is drawing white background, before it was drawing ok but failing because of line width.

Before:
Before

After:
After

**Review by louwers — CHANGES_REQUESTED** [review]
Should we make the more correct (less performant) version the default?

@sjg-wdw @stefankarschti Can one of you create an issue that links to this PR to re-address this after variants have been introduced?

**Comment by stefankarschti** [comment]
While investigating the failed render tests I realized that we also need shader variations for the triangulated lines in case of outline-color and opacity per-vertex, which is the case with some expressions. 
`render-tests/fill-outline-color/zoom-and-property-function`
`render-tests/fill-outline-color/property-function`
ref #1855

**Pull request closed without merging**